File: src/main/kotlin/ru/arcam/yggdrasil/service/DockerController.kt
package ru.arcam.yggdrasil.service
import ru.arcam.yggdrasil.leaf.Leaf
import java.io.BufferedReader
import java.io.InputStreamReader

class DockerController(leaf: Leaf) : IController(leaf) {
    override fun stop(): String {
        val SERVER_STATUS = arrayOf("docker", "stop", leaf.name)
        try {
            val process = ProcessBuilder(*SERVER_STATUS).redirectErrorStream(true).start()
            process.waitFor()
            return "OK"
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return "ERROR"
    }
    
    override fun start(): String {
        val SERVER_STATUS = arrayOf("docker", "start", leaf.name)
        try {
            val process = ProcessBuilder(*SERVER_STATUS).redirectErrorStream(true).start()
            process.waitFor()
            return "OK"
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return "ERROR"
    }

    override fun restart(): String {
        val SERVER_STATUS = arrayOf("docker", "restart", leaf.name)
        try {
            val process = ProcessBuilder(*SERVER_STATUS).redirectErrorStream(true).start()
            process.waitFor()
            return "OK"
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return "ERROR"
    }
    
    override fun status(): String {
        val SERVER_STATUS = arrayOf("docker", "ps", "-f", "name=" + leaf.name)
        try {
            val process = ProcessBuilder(*SERVER_STATUS).redirectErrorStream(true).start()
            process.waitFor()
            val input = process.inputStream
            val br = BufferedReader(InputStreamReader(input))
            var line: String?
            while (br.readLine().also { line = it } != null) {
                if (line!!.contains("Up"))
                    return "RUNNING"
            }
            return "STOPPED"
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return "UNAVAIVABLE"
    }

    override fun logs(args: List<String>?): String {
        var num = 10
        if (args != null && args.size > 0) {
            val tmp = args[0].toIntOrNull()
            if (tmp != null && tmp > 0)
                num = tmp
        }
        val SERVER_LOGS = arrayOf("docker", "logs", "--name", leaf.name, "-n", num.toString())
        try {
            val process = ProcessBuilder(*SERVER_LOGS).redirectErrorStream(true).start()
            process.waitFor()
            val input = process.inputStream
            val br = BufferedReader(InputStreamReader(input))
            val result = br.readLines().joinToString { it -> "\n" + it }
            return result
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return "UNAVAIVABLE"
    }

    override fun callMethod(method: String, args: List<String>): String {
        return "OK"
    }
}

File: service/docker_controller.go
package service

import (
	"bufio"
	"os/exec"
	"strings"
	"yggdrasil-branch/models"
)

// DockerController управляет Docker контейнерами
type DockerController struct {
	leaf *models.Leaf
}

// NewDockerController создает новый DockerController
func NewDockerController(leaf *models.Leaf) *DockerController {
	return &DockerController{leaf: leaf}
}

// GetLeaf возвращает лист
func (dc *DockerController) GetLeaf() *models.Leaf {
	return dc.leaf
}

// Stop останавливает контейнер
func (dc *DockerController) Stop() string {
	cmd := exec.Command("docker", "stop", dc.leaf.Name)
	if err := cmd.Run(); err != nil {
		return "ERROR"
	}
	return "OK"
}

// Start запускает контейнер
func (dc *DockerController) Start() string {
	cmd := exec.Command("docker", "start", dc.leaf.Name)
	if err := cmd.Run(); err != nil {
		return "ERROR"
	}
	return "OK"
}

// Restart перезапускает контейнер
func (dc *DockerController) Restart() string {
	cmd := exec.Command("docker", "restart", dc.leaf.Name)
	if err := cmd.Run(); err != nil {
		return "ERROR"
	}
	return "OK"
}

// Status возвращает статус контейнера
func (dc *DockerController) Status() string {
	cmd := exec.Command("docker", "ps", "-f", "name="+dc.leaf.Name)
	output, err := cmd.Output()
	if err != nil {
		return "UNAVAIVABLE"
	}
	
	scanner := bufio.NewScanner(strings.NewReader(string(output)))
	for scanner.Scan() {
		line := scanner.Text()
		if strings.Contains(line, "Up") {
			return "RUNNING"
		}
	}
	return "STOPPED"
}

// Logs возвращает логи контейнера
func (dc *DockerController) Logs(args []string) string {
	num := 10
	if len(args) > 0 {
		if n, err := parseInt(args[0]); err == nil && n > 0 {
			num = n
		}
	}
	
	cmd := exec.Command("docker", "logs", dc.leaf.Name, "-n", intToString(num))
	output, err := cmd.Output()
	if err != nil {
		return "UNAVAIVABLE"
	}
	return string(output)
}

// CallMethod вызывает произвольный метод
func (dc *DockerController) CallMethod(method string, args []string) string {
	return "OK"
}
